File: Spring/les/src/main/java/com/jga/les/model/Compra.java
package com.jga.les.model;

import java.util.Date;
import java.util.List;

import com.fasterxml.jackson.annotation.JsonIgnore;
import org.hibernate.annotations.Cascade;
import org.hibernate.annotations.CascadeType;
import org.hibernate.annotations.CreationTimestamp;
import org.springframework.format.annotation.DateTimeFormat;

import com.fasterxml.jackson.annotation.JsonManagedReference;

import io.micrometer.common.lang.NonNull;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import lombok.Data;

@Data
@Entity
public class Compra {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;

    @NonNull
    @DateTimeFormat(pattern = "dd/MM/yyyy")
    @CreationTimestamp
    private Date entrada;

    private Date saida;

    @ManyToOne
    private Cliente cliente;

    @OneToMany(mappedBy = "compra", orphanRemoval = true)
    @Cascade(CascadeType.ALL)
    @JsonManagedReference
    private List<CompraProduto> compraProdutos;

    @JsonIgnore
    public double getTotal(){
        double total = 0;
        for (CompraProduto compraProduto : compraProdutos){
            total += compraProduto.getPreco() * compraProduto.getQntd();
        }
        return total;
    }
}

File: Flutter/les/lib/model/compra/compra.dart
import 'package:les/model/cliente/cliente.dart';
import 'package:les/model/compra/compra_produto.dart';
import 'package:les/model/entity.dart';

class Compra extends Entity{
  DateTime entrada;
  DateTime? saida;
  Cliente cliente;
  List<CompraProduto> compraProdutos;

  Compra({
    super.id,
    required this.entrada,
    this.saida,
    required this.cliente,
    required this.compraProdutos,
  });

  factory Compra.fromJson(Map<String, dynamic> json){
    return Compra(
      id: json['id'],
      entrada: DateTime.parse(json['entrada']),
      saida: json["saida"] == null ? null : DateTime.parse(json['saida']),
      cliente: Cliente.fromJson(json['cliente']),
      compraProdutos: json['compraProdutos'] == null ? [] : (json['compraProdutos'] as List).map((produto) => CompraProduto.fromJson(produto)).toList(),
    );
  }

  @override
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'entrada': entrada.toIso8601String(),
      'saida': saida?.toIso8601String(),
      'cliente': cliente.toJson(),
      'compraProdutos': compraProdutos.map((produto) => produto.toJson()).toList(),
    };
  }
}